token-collections/program/src/entrypoint.rs

```diff
@@ -0,0 +1,23 @@
+//! Program entrypoint
+
+use {
+    crate::processor,
+    solana_program::{
+        account_info::AccountInfo, entrypoint, entrypoint::ProgramResult,
+        program_error::PrintProgramError, pubkey::Pubkey,
+    },
+    spl_token_group_interface::error::TokenGroupError,
+};
+
+entrypoint!(process_instruction);
+fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    if let Err(error) = processor::process(program_id, accounts, instruction_data) {
+        error.print::<TokenGroupError>();
+        return Err(error);
+    }
```

Comment:
I was actually always curious why you did this. Aren't we converting all of our errors to `ProgramError` in each `Result`, and doesn't the runtime automatically print `ProgramError::Custom(..)` ?

Reply:
yes, but `PrintProgramError` allows us to specify one error type, and if it's `ProgramError::Custom`, it'll use that error's special printer. so if you do:

```
        let error = ProgramError::Custom(TokenError::OwnerMismatch as u32);
        error.print::<TokenError>();
```

You'll get

```
Error: owner does not match
```

And if it's outside the range, you just get `Error: Unknown`. So probably what we want is some way to declare *multiple* possible error types and then use the appropriate printer depending on the `ProgramError::Custom` returned, but that's not very easy... anyway, a problem for another day!

Reply:
Ohhh this is actually a pretty big problem considering it will attempt to match the `u32` value and you can often get the wrong error printed. Yikes.

Reply:
I'm not really sure this is possible to do with the way we're doing `Result<_, ProgramError>`, since the functions are always going to return `ProgramError::Custom(u32)` for any custom errors. There's no way right now to know which type of error this came from.

The only not-insane way to do this within only SPL is to rely _heavily_ on the "hashing" changes I pushed to the SPL errors a bit ago (#5169). We'd need to add the `From<ProgramError>` trait to each error as well. Then it should be OK as long as we're totally sure none of them share the same codes (eek).

Reply:
Yeah exactly. We might need to change processor functions to return some other printable type, like `std::error::Error`, which `thiserror` gives us, and then print that when it's returned. It has higher compute usage, but it's probably not a huge price to pay for correct error strings.

Anyway, not a problem that's super relevant here, but one to think about!

Reply:
#5660